sensor/common/sensor/buffered_stream.go

```diff
@@ -0,0 +1,65 @@
+package sensor
+
+import (
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/sensor/common/messagestream"
+	"github.com/stackrox/rox/sensor/common/metrics"
+)
+
+type bufferedStream struct {
+	buffer chan *central.MsgFromSensor
+	stopC  concurrency.ReadOnlyErrorSignal
+	stream messagestream.SensorMessageStream
+}
+
+func (s bufferedStream) Send(msg *central.MsgFromSensor) error {
+	select {
+	case s.buffer <- msg:
+		metrics.ResponsesChannelAdd()
+	default:
+		// The buffer is full, we drop the message and return
+		metrics.IncResponsesChannelDroppedCount(msg)
+		return nil
+	}
+	return nil
+}
+
+func NewBufferedStream(stream messagestream.SensorMessageStream, msgC chan *central.MsgFromSensor, stopC concurrency.ReadOnlyErrorSignal) (messagestream.SensorMessageStream, <-chan error) {
+	// if the capacity of the buffer is zero then we just return the inner stream
+	if cap(msgC) == 0 {
+		return stream, nil
+	}
+	ret := bufferedStream{
+		buffer: msgC,
+		stopC:  stopC,
+		stream: stream,
+	}
+	errC := ret.run()
+	return ret, errC
+}
+
+func (s bufferedStream) run() <-chan error {
+	errC := make(chan error)
```

Comment:
I have not fully investigated possibilities of this channel blocking, but shall we maybe buffer it too?
Not sure if the same size as in  `ROX_RESPONSES_CHANNEL_BUFFER_SIZE` is required, but provided the context in which we are implementing this fix (i.e. competing for channels and locking resources), maybe it makes sense to add a buffer that could be a, for example, `math.Min(100,env.ResponsesChannelBufferSize.IntegerSetting())`. WDYT?

Reply:
🤔 I'm not sure if we want to do that. `errC` will block but just like the previous implementation. The idea is that if `Send` blocks we will start buffering in `buffer` until we reach the limit. If we make `errC` buffered we will have two buffers and we could potentially drop from both?

Reply:
I was thinking about a simple buffer without counting the drops - simply `make(chan error, 100)` to prevent potential blocking in a place where we expect buffering.
I am not insisting on this. If the reader of the `errC` is simple and has low chance of causing problems, then let's scratch that idea.

Reply:
Leaving this comment open - fell free to resolve after you decide what to do here (noop is fine 😉 ).